url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/grammar/RuleGrammar.java

```diff
@@ -0,0 +1,244 @@
+		}
+		return alt;
+	}
+
+	private Symbol sym(String symbol) {
+		return new Symbol(symbol);
+	}
+
+	class SequenceAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			Parser last = (Parser) a.pop();
+			Parser butlast = (Parser) a.pop();
+			Sequence seq;
+			if (last instanceof Sequence) {
+				seq = (Sequence) last;
+			} else {
+				seq = seq(last);
+			}
+			seq.addTop(butlast);
+			a.push(seq);
+		}
+	}
+
+	class AlternationAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			Parser last = (Parser) a.pop();
+			Parser butlast = (Parser) a.pop();
+			Alternation alt;
+			if (butlast instanceof Alternation) {
+				alt = (Alternation) butlast;
+			} else {
+				alt = alt(butlast);
+			}
+			alt.add(last);
+			a.push(alt);
+		}
+	}
+
+	class RuleAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			if (a.elementsRemaining() > 0)
+				return;
+			Parser parser = (Parser) a.pop();
+			String ruleName = ((Token) a.pop()).sval();
+			targetGrammar.addRule(ruleName, parser);
+			a.push(ruleName);
+		}
+	}
+
+	class CaselessAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			String quoted = ((Token) a.pop()).sval();
+			String caseless = quoted.substring(1, quoted.length() - 1);
+			CaselessLiteral constant = new CaselessLiteral(caseless);
+			if (targetGrammar.areAllConstantsDiscarded())
+				constant.discard();
+			a.push(constant);
+		}
+	}
+
+	class DiscardAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			Terminal t = (Terminal) a.getStack().peek();
+			t.discard();
+		}
+	}
+
+	class SymbolAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			String quoted = ((Token) a.pop()).sval();
+			String symbol = quoted.substring(1, quoted.length() - 1);
+			Symbol constant = new Symbol(symbol);
+			if (targetGrammar.areAllConstantsDiscarded())
+				constant.discard();
+			a.push(constant);
+		}
+	}
+
+	class TerminalAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			String terminalType = ((Token) a.pop()).sval();
+			a.push(targetGrammar.terminal(terminalType));
+		}
+	}
+
+	class ReferenceAssembler implements IAssembler {
+		public void workOn(Assembly a) {
+			String ruleName = ((Token) a.pop()).sval();
+			a.push(new RuleReference(ruleName, targetGrammar));
+		}
+	}
+
+	class RepetitionAssembler implements IAssembler {
+		private final int requiredMatches;
+
+		public RepetitionAssembler(int requiredMatches) {
+			this.requiredMatches = requiredMatches;
+		}
+
+		public void workOn(Assembly a) {
+			Parser atomicStep = (Parser) a.pop();
+			a.push(new Repetition(atomicStep).requireMatches(requiredMatches));
+		}
+	}
+}
+
+class CaselessLiteralDef extends QuotedString {
+	@Override
+	protected boolean qualifies(Object o) {
+		if (!super.qualifies(o))
+			return false;
+		Token t = (Token) o;
+		String quotedString = t.sval();
+		if (!quotedString.matches("\"\\w+\""))
+			return false;
+		return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 201:207, 225:231.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.